File: framework/dotnet/BetterPortal/Registry.cs
using AnyVali;
using System.Collections.Frozen;
using System.Text.RegularExpressions;
using Node = System.Collections.Generic.Dictionary<string, object?>;

namespace BetterPortal;

public sealed class Operation
{
    private readonly Node declaration;
    public Handler Handler { get; }
    public IReadOnlyList<Renderer<Generated.ViewRenderError>> ErrorRenderers { get; }
    public string Id => (string)declaration["operationId"]!;
    public string Method => (string)declaration["method"]!;
    public Generated.OperationDeclaration Declaration => Contracts.Parse<Generated.OperationDeclaration>("OperationDeclarationSchema", declaration);
    public Operation(Handler handler, Generated.OperationDeclarationInput declaration, IEnumerable<Renderer<Generated.ViewRenderError>>? errorRenderers = null)
    {
        Handler = handler; this.declaration = (Node)Contracts.Parse("OperationDeclarationSchema", declaration)!;
        if (handler.IsStreaming && Method != "GET") throw new ArgumentException("Finite streams require a GET operation");
        ErrorRenderers = Renderer.Unique(errorRenderers ?? []);
        if (ErrorRenderers.Any(item => item.Identity.Status < 400)) throw new ArgumentException("Error renderers require an error status");
    }
    public ValueTask<object?> Invoke(RequestContext context, IReadOnlyDictionary<string, object?> values, CancellationToken cancellation = default)
    {
        if (context.Method != Method) throw new ArgumentException("Operation method does not match the request context");
        return Handler.InvokeBoxed(context, values, cancellation);
    }
    public ValueTask<Invocation> Execute(RequestContext context, IReadOnlyDictionary<string, object?> values, CancellationToken cancellation = default)
    {
        if (context.Method != Method) throw new ArgumentException("Operation method does not match the request context");
        return Handler.ExecuteBoxed(context, values, cancellation);
    }
    public async ValueTask<RawResponse?> RenderError(RenderContext context, string message)
    {
        var route = context.Route; var kind = route.Kind.ToString().ToLowerInvariant();
        var identity = (route.Renderer, kind, route.Key.HasValue ? route.Key.Value : null, (int)route.Status);
        var renderer = ErrorRenderers.FirstOrDefault(item => item.Identity == identity);
        if (renderer is null && !Handler.Renderers.Any(item => (item.Identity.Renderer, item.Identity.Kind, item.Identity.Key) == (identity.Renderer, kind, identity.Item3))) return null;
        var html = renderer is null ? "" : await renderer.Render(Contracts.Parse<Generated.ViewRenderError>("ViewRenderErrorSchema", new { error = message, status = route.Status }), context);
        var mode = kind == "page" ? route.Mode.ToString().ToLowerInvariant() : "fragment";
        return new RawResponse(System.Text.Encoding.UTF8.GetBytes(html), (int)route.Status, new Dictionary<string, string> { ["content-type"] = HtmlContentType(mode) });
    }
    public string HtmlContentType(string mode) => Renderer.ContentType(mode, declaration.GetValueOrDefault("chrome"));
    internal static Node Export(Schema schema) => (Node)Json.Read(Json.Write(V.Export(schema)))!;
    internal Node Metadata(string viewId, IReadOnlyDictionary<string, string> aliases, string pluginId)
    {
        var result = (Node)Json.Read(Json.Write(declaration))!;
        foreach (var (source, target) in new[] { ("query", "querySchema"), ("headers", "headersSchema"), ("request", "bodySchema") })
            result[target] = Handler.Schemas.TryGetValue(source, out var schema) ? Export(schema) : new Node();
        result["jsonResponseSchema"] = Handler.ResponseSchema is { } response ? Export(response) : new Node(); result["metadataResponseSchema"] = new Node();
        if (Handler.IsRaw) result["raw"] = true;
        if (Handler.StreamingMetadata is { } streaming) result["streaming"] = streaming;
        var html = Renderer.HtmlMetadata(Handler.Renderers, Handler.StreamRendererKeys);
        result["renderable"] = ((Node)html["renderers"]!).Count > 0; result["html"] = html;
        result.TryAdd("sitemap", new Node { ["kind"] = "default" });
        foreach (var dependency in ((List<object?>)result["dependencies"]!).Cast<Node>())
        {
            if (!dependency.TryGetValue("serviceId", out var alias)) continue;
            if (!aliases.TryGetValue((string)alias!, out var service)) throw new ArgumentException("Unknown dependency alias: " + alias);
            if (service == pluginId) dependency.Remove("serviceId"); else dependency["serviceId"] = service;
        }
        var callers = ((List<object?>)((Node)result["auth"]!)["callers"]!).Cast<string>().ToHashSet(StringComparer.Ordinal);
        foreach (var descriptor in ((List<object?>)result["apiContracts"]!).Cast<Node>())
        {
            if (!((List<object?>)descriptor["modes"]!).Cast<string>().All(callers.Contains))
                throw new ArgumentException("API contract caller modes are not allowed by the operation");
            descriptor["viewId"] = viewId; descriptor["methods"] = new List<object?> { Method };
        }
        foreach (var rule in ((List<object?>)result["robots"]!).Cast<Node>())
            if (!Regex.IsMatch((string)rule["userAgent"]!, @"\A[A-Za-z0-9*._-]{1,100}\z")) throw new ArgumentException("Invalid robots user-agent token");
        return (Node)Contracts.Parse("ViewOperationMetadataSchema", result)!;
    }
}

public sealed class Route
{
    public string ViewId { get; }
    public IReadOnlyList<string> Paths { get; }
    public IReadOnlyList<Operation> Operations { get; }
    public SseFeed? Sse { get; }
    public IReadOnlyList<string> ParamNames => Paths.SelectMany(path => Segments(path).Where(part => part.StartsWith(':')).Select(part => part[1..])).Distinct(StringComparer.Ordinal).ToArray();
    public Route(string viewId, string path, IEnumerable<Operation> operations, IEnumerable<string>? pathVariants = null, SseFeed? sse = null)
    {
        ViewId = (string)Contracts.Parse(Contracts.Get("ViewMetadataSchema", "viewId"), viewId)!;
        Paths = Array.AsReadOnly(new[] { path }.Concat(pathVariants ?? []).Distinct(StringComparer.Ordinal)
            .OrderByDescending(value => Segments(value).Count(part => part.StartsWith(':'))).ThenByDescending(value => value.Length).ThenBy(value => value, StringComparer.Ordinal).ToArray());
        Operations = Array.AsReadOnly(operations.ToArray());
        Sse = sse;
        if (Operations.Count == 0 || Operations.Select(operation => operation.Method).Distinct(StringComparer.Ordinal).Count() != Operations.Count)
            throw new ArgumentException("A route needs unique method operations");
        if (sse is not null && (sse.ViewId != ViewId || !Operations.Any(operation => operation.Method == "GET" && ReferenceEquals(operation.Handler, sse.Owner))))
            throw new ArgumentException("SSE must belong to this view's GET handler");
        var schemas = Operations.Select(operation => operation.Handler.Schemas.TryGetValue("params", out var schema) ? Operation.Export(schema) : new Node()).ToArray();
        if (schemas.Skip(1).Any(schema => !System.Text.Json.Nodes.JsonNode.DeepEquals(System.Text.Json.Nodes.JsonNode.Parse(Json.Write(schemas[0])),
            System.Text.Json.Nodes.JsonNode.Parse(Json.Write(schema))))) throw new ArgumentException("A view must publish one consistent params schema");
    }
    internal static string[] Segments(string path)
    {
        if (!path.StartsWith('/') || path.IndexOfAny(['?', '#', '\\', '%']) >= 0 || path.Any(c => c < 33)) throw new ArgumentException("Expected a root-relative route path");
        var segments = path == "/" ? [] : path[1..].Split('/');
        if (segments.Any(part => part.Length == 0 || part is "." or ".." || part.IndexOfAny(['[', ']', '{', '}']) >= 0)) throw new ArgumentException("Invalid route segment");
        var parameters = segments.Where(part => part.StartsWith(':')).Select(part => part[1..]).ToArray();
        if (parameters.Any(name => !Regex.IsMatch(name, @"\A[A-Za-z_][A-Za-z0-9_]*\z")) || parameters.Distinct(StringComparer.Ordinal).Count() != parameters.Length)
            throw new ArgumentException("Invalid or duplicate route parameter");
        if (segments.Any(part => part.Contains(':') && !part.StartsWith(':'))) throw new ArgumentException("Parameters must occupy a complete route segment");
        return segments;
    }
}

public sealed class Registry
{
    public IReadOnlyList<Route> Routes { get; }
    public IReadOnlyDictionary<string, string> Dependencies { get; }
    public Registry(IEnumerable<Route> routes, IReadOnlyDictionary<string, string>? dependencies = null)
    {
        Routes = Array.AsReadOnly(routes.ToArray());
        Dependencies = (dependencies ?? new Dictionary<string, string>()).ToFrozenDictionary(pair => pair.Key,
            pair => (string)Contracts.Parse("PluginIdSchema", pair.Value)!, StringComparer.Ordinal);
        var viewIds = new HashSet<string>(StringComparer.Ordinal); var operationIds = new HashSet<string>(StringComparer.Ordinal); var paths = new HashSet<(string, string)>();
        foreach (var route in Routes)
        {
            if (!viewIds.Add(route.ViewId)) throw new ArgumentException("Duplicate view ID: " + route.ViewId);
            foreach (var operation in route.Operations)
            {
                if (!operationIds.Add(operation.Id)) throw new ArgumentException("Duplicate operation ID: " + operation.Id);
                foreach (var path in route.Paths)
                    if (!paths.Add((operation.Method, string.Join('/', Route.Segments(path).Select(part => part.StartsWith(':') ? ":" : part)))))
                        throw new ArgumentException("Ambiguous route path: " + path);
            }
        }
    }
    private Node ManifestNode(Generated.ManifestDeclarationInput declaration)
    {
        var result = (Node)Contracts.Parse("ManifestDeclarationSchema", declaration)!;
        var pluginId = (string)result["pluginId"]!;
        var local = Routes.SelectMany(route => route.Operations.Select(operation => (operation.Id, operation.Method))).ToHashSet();
        var views = new List<object?>(); var contracts = new List<object?>((List<object?>)result["apiContracts"]!);
        foreach (var route in Routes)
        {
            var operations = route.Operations.Select(operation => operation.Metadata(route.ViewId, Dependencies, pluginId)).ToArray();
            foreach (var operation in operations)
            {
                foreach (var dependency in ((List<object?>)operation["dependencies"]!).Cast<Node>())
                    if (!dependency.ContainsKey("serviceId") && !local.Contains(((string)dependency["operationId"]!, (string)dependency["method"]!)))
                        throw new ArgumentException("Unavailable local operation dependency");
                contracts.AddRange((List<object?>)operation["apiContracts"]!);
            }
            var primary = operations.FirstOrDefault(operation => operation["method"] is "GET") ?? operations[0];
            views.Add(new Node { ["viewId"] = route.ViewId, ["path"] = route.Paths[0], ["pathVariants"] = route.Paths.Count > 1 ? route.Paths : [],
                ["title"] = primary["title"], ["description"] = primary["description"], ["operations"] = operations,
                ["paramsSchema"] = route.Operations[0].Handler.Schemas.TryGetValue("params", out var schema) ? Operation.Export(schema) : new Node() });
        }
        var capabilities = ((List<object?>)result["capabilities"]!).Cast<string>().Concat(["view.json", "view.metadata"]).ToList();
        var renderers = new List<string>(); var modes = new List<string>();
        if (result.TryGetValue("shell", out var shell)) { renderers.Add((string)((Node)shell!)["renderer"]!); capabilities.Add("renderer." + renderers[0]); }
        foreach (var handler in Routes.SelectMany(route => route.Operations).Select(operation => operation.Handler).Where(handler => handler.IsStreaming))
        {
            capabilities.Add("stream.ndjson");
            foreach (var theme in handler.StreamRendererKeys)
            {
                renderers.Add(theme); modes.Add("fragment"); capabilities.AddRange(["renderer." + theme, "view.sse-render", "view.html"]);
            }
        }
        foreach (var renderer in Routes.SelectMany(route => route.Operations).SelectMany(operation => operation.Handler.Renderers))
        {
            var (theme, kind, _, status) = renderer.Identity;
            if (status != 200) continue;
            renderers.Add(theme); capabilities.Add("renderer." + theme);
            if (kind == "page") modes.Add("page");
            if (kind == "fragment") modes.Add("fragment");
            if (kind is "page" or "component") capabilities.Add("view.html");
        }
        if (((List<object?>)result["configSchemas"]!).Count > 0)
        {
            var apis = (List<object?>)result["adminApis"]!;
            var existing = apis.Cast<Node>().Select(api => (string)api["id"]!).ToHashSet(StringComparer.Ordinal);
            foreach (var (id, title, description, path, methods) in new[] {
                ("config.schema", "Config Schema", "BetterPortal-managed config schemas for this service.", "/.well-known/bp/config/schema", new[] { "GET" }),
                ("config.values", "Config Values", "Read and write BetterPortal-managed config values.", "/.well-known/bp/config", new[] { "GET", "POST" }) })
                if (!existing.Contains(id)) apis.Add(new Node { ["id"] = id, ["title"] = title, ["description"] = description, ["path"] = path, ["methods"] = methods, ["supportsCustomUi"] = false });
        }
        result["protocolVersion"] = 2; result["views"] = views; result["capabilities"] = capabilities.Distinct(StringComparer.Ordinal).ToArray();
        result["supportedRenderers"] = renderers.Distinct().ToArray(); result["supportedRenderModes"] = modes.Distinct().ToArray(); result["apiContracts"] = contracts;
        return (Node)Contracts.Parse("PluginManifestSchema", result)!;
    }
    public Generated.PluginManifest Manifest(Generated.ManifestDeclarationInput declaration) => Contracts.Parse<Generated.PluginManifest>("PluginManifestSchema", ManifestNode(declaration));
    public Generated.BpSchemaOutput Schema(Generated.ManifestDeclarationInput declaration) => Contracts.Parse<Generated.BpSchemaOutput>("BpSchemaOutputSchema", new Node
    {
        ["manifest"] = ManifestNode(declaration), ["routes"] = Routes.Select(route => new Node
        {
            ["viewId"] = route.ViewId, ["path"] = route.Paths[0], ["pathVariants"] = route.Paths.Count > 1 ? route.Paths : [],
            ["operations"] = route.Operations.Select(operation => new { operationId = operation.Id, method = operation.Method }).ToArray(),
            ["paramNames"] = route.ParamNames, ["renderers"] = RouteRenderers(route).Select(item => item.Renderer.Identity.Renderer)
                .Concat(route.Operations.SelectMany(operation => operation.Handler.StreamRendererKeys)).Distinct().ToArray(),
            ["hasFragments"] = RouteRenderers(route).Any(item => item.Renderer.Identity.Kind == "fragment"),
            ["fragments"] = RouteRenderers(route).Where(item => item.Renderer.Identity.Kind == "fragment")
                .GroupBy(item => (item.Renderer.Identity.Key, item.Operation.Id, item.Operation.Method)).Select(group => new Node {
                    ["fragmentLocation"] = group.Key.Key!.Split('.')[0], ["fragmentId"] = group.Key.Key!.Split('.')[1], ["operationId"] = group.Key.Id,
                    ["method"] = group.Key.Method, ["renderers"] = group.Select(item => item.Renderer.Identity.Renderer).ToArray() }).ToArray(),
            ["components"] = RouteRenderers(route).Where(item => item.Renderer.Identity.Kind == "component").Select(item => item.Renderer.Identity.Key).Distinct().ToArray()
        }).ToArray()
    });
    private static IEnumerable<(Operation Operation, Renderer Renderer)> RouteRenderers(Route route) => route.Operations.SelectMany(operation =>
        operation.Handler.Renderers.Where(renderer => renderer.Identity.Status == 200).Select(renderer => (operation, renderer)));
}

File: framework/python/betterportal/registry.py
"""Operation registration and manifests derived from native AnyVali contracts."""
from __future__ import annotations

from copy import deepcopy
import re
from types import MappingProxyType
from typing import Any, Iterable, Mapping, cast

from .contracts import contract, export, parse
from .generated_types import OperationDeclarationInput, OperationDeclaration, ManifestDeclarationInput, PluginManifest, BpSchemaOutput, ViewRenderError
from .handler import Handler, RequestContext, Invocation
from .response import RawHandler, RawResponse
from .rendering import Renderer, RenderContext, html_metadata, content_type, renderers as unique_renderers


class Operation:
    def __init__(self, handler: Handler[Any, Any, Any, Any, Any] | RawHandler[Any, Any, Any, Any], declaration: OperationDeclarationInput, *, error_renderers: Iterable[Renderer[ViewRenderError]] = ()):
        self.handler = handler
        self._declaration = parse("OperationDeclarationSchema", declaration)
        self.error_renderers = unique_renderers(error_renderers)
        if any(item.identity[3] < 400 for item in self.error_renderers): raise ValueError("Error renderers require an error status")

    @property
    def declaration(self) -> OperationDeclaration:
        return cast(OperationDeclaration, deepcopy(self._declaration))

    @property
    def id(self) -> str: return self._declaration["operationId"]

    @property
    def method(self) -> str: return self._declaration["method"]

    async def invoke(self, context: RequestContext, values: Mapping[str, Any]) -> Any:
        return (await self.execute(context, values)).value

    async def execute(self, context: RequestContext, values: Mapping[str, Any]) -> Invocation[Any]:
        if context.method != self.method:
            raise ValueError("Operation method does not match the request context")
        return await self.handler.execute(context, values)

    async def render_error(self, context: RenderContext, message: str) -> RawResponse | None:
        route = context.route
        identity = (route["renderer"], route["kind"], route.get("key"), route["status"])
        renderer = next((item for item in self.error_renderers if item.identity == identity), None)
        if renderer is None and not any(item.identity[:3] == identity[:3] for item in self.handler.renderers): return None
        html = await renderer.render(parse("ViewRenderErrorSchema", {"error": message, "status": route["status"]}), context) if renderer is not None else ""
        mode = route["mode"] if route["kind"] == "page" else "fragment"
        return RawResponse(html.encode("utf-8"), status=route["status"], headers={"content-type": content_type(mode, self._declaration.get("chrome"))})

    def metadata(self, view_id: str, aliases: Mapping[str, str], plugin_id: str) -> dict[str, Any]:
        result = deepcopy(self._declaration)
        result.update({target: export(self.handler.schemas[source]) if source in self.handler.schemas else {}
                       for source, target in (("query", "querySchema"), ("headers", "headersSchema"), ("request", "bodySchema"))})
        html = html_metadata(self.handler.renderers)
        result.update(jsonResponseSchema=export(self.handler.response_schema) if self.handler.response_schema is not None else {}, metadataResponseSchema={}, renderable=bool(html["renderers"]), html=html)
        if self.handler.is_raw: result["raw"] = True
        result.setdefault("sitemap", {"kind": "default"})
        for dependency in result["dependencies"]:
            if "serviceId" in dependency:
                alias = dependency["serviceId"]
                if alias not in aliases:
                    raise ValueError("Unknown dependency alias: " + alias)
                if aliases[alias] == plugin_id: dependency.pop("serviceId")
                else: dependency["serviceId"] = aliases[alias]
        for descriptor in result["apiContracts"]:
            if not set(descriptor.get("modes", ["service"])).issubset(result["auth"]["callers"]):
                raise ValueError("API contract caller modes are not allowed by the operation")
            descriptor.update(viewId=view_id, methods=[self.method])
        for rule in result["robots"]:
            if not re.fullmatch(r"[A-Za-z0-9*._-]{1,100}", rule["userAgent"]):
                raise ValueError("Invalid robots user-agent token")
        return parse("ViewOperationMetadataSchema", result)


def _segments(path: str) -> tuple[str, ...]:
    if not path.startswith("/") or any(char in path for char in "?#\\%") or any(ord(char) < 33 for char in path):
        raise ValueError("Expected a root-relative route path")
    segments = tuple(path[1:].split("/")) if path != "/" else ()
    if any(not segment or segment in (".", "..") or any(char in segment for char in "[]{}") for segment in segments):
        raise ValueError("Invalid route segment")
    parameters = [segment[1:] for segment in segments if segment.startswith(":")]
    if any(not re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", name) for name in parameters) or len(parameters) != len(set(parameters)):
        raise ValueError("Invalid or duplicate route parameter")
    if any(":" in segment and not segment.startswith(":") for segment in segments):
        raise ValueError("Parameters must occupy a complete route segment")
    return segments


class Route:
    def __init__(self, view_id: str, path: str, operations: Iterable[Operation], *, path_variants: Iterable[str] = ()):
        self.view_id = contract("ViewMetadataSchema", "viewId").parse(view_id)
        self.paths = tuple(sorted(set([path, *path_variants]), key=lambda value: (-sum(segment.startswith(":") for segment in _segments(value)), -len(value), value)))
        self.operations = tuple(operations)
        if not self.operations or len({operation.method for operation in self.operations}) != len(self.operations):
            raise ValueError("A route needs unique method operations")
        schemas = [export(operation.handler.schemas["params"]) if "params" in operation.handler.schemas else {} for operation in self.operations]
        if any(schema != schemas[0] for schema in schemas[1:]):
            raise ValueError("A view must publish one consistent params schema")

    @property
    def param_names(self) -> list[str]:
        return list(dict.fromkeys(segment[1:] for path in self.paths for segment in _segments(path) if segment.startswith(":")))


class Registry:
    def __init__(self, routes: Iterable[Route], *, dependencies: Mapping[str, str] | None = None):
        self.routes = tuple(routes)
        self.dependencies = MappingProxyType({alias: parse("PluginIdSchema", plugin) for alias, plugin in (dependencies or {}).items()})
        view_ids, operation_ids, paths = set(), set(), set()
        for route in self.routes:
            if route.view_id in view_ids: raise ValueError("Duplicate view ID: " + route.view_id)
            view_ids.add(route.view_id)
            for operation in route.operations:
                if operation.id in operation_ids: raise ValueError("Duplicate operation ID: " + operation.id)
                operation_ids.add(operation.id)
                for path in route.paths:
                    key = (operation.method, tuple(":" if part.startswith(":") else part for part in _segments(path)))
                    if key in paths: raise ValueError("Ambiguous route path: " + path)
                    paths.add(key)

    def manifest(self, declaration: ManifestDeclarationInput) -> PluginManifest:
        result = parse("ManifestDeclarationSchema", declaration)
        plugin_id = result["pluginId"]
        local = {(operation.id, operation.method) for route in self.routes for operation in route.operations}
        views, contracts = [], list(result["apiContracts"])
        for route in self.routes:
            operations = [operation.metadata(route.view_id, self.dependencies, plugin_id) for operation in route.operations]
            for operation in operations:
                for dependency in operation["dependencies"]:
                    if "serviceId" not in dependency and (dependency["operationId"], dependency["method"]) not in local:
                        raise ValueError("Unavailable local operation dependency")
                contracts.extend(operation["apiContracts"])
            primary = next((item for item in operations if item["method"] == "GET"), operations[0])
            schema = route.operations[0].handler.schemas.get("params")
            views.append({"viewId": route.view_id, "path": route.paths[0], "pathVariants": list(route.paths) if len(route.paths) > 1 else [],
                          "title": primary["title"], "description": primary["description"], "paramsSchema": export(schema) if schema is not None else {}, "operations": operations})
        capabilities = list(dict.fromkeys([*result["capabilities"], "view.json", "view.metadata"]))
        renderers, modes = [], []
        if "shell" in result:
            renderers.append(result["shell"]["renderer"])
            capabilities.append("renderer." + renderers[0])
        for route in self.routes:
            for registered in route.operations:
                for renderer in registered.handler.renderers:
                    theme, kind, _, status = renderer.identity
                    if status != 200: continue
                    renderers.append(theme); capabilities.append("renderer." + theme)
                    if kind == "page": modes.append("page")
                    if kind == "fragment": modes.append("fragment")
                    if kind in ("page", "component"): capabilities.append("view.html")
        if result["configSchemas"]:
            existing = {item["id"] for item in result["adminApis"]}
            for identifier, title, description, path, methods in (
                ("config.schema", "Config Schema", "BetterPortal-managed config schemas for this service.", "/.well-known/bp/config/schema", ["GET"]),
                ("config.values", "Config Values", "Read and write BetterPortal-managed config values.", "/.well-known/bp/config", ["GET", "POST"])):
                if identifier not in existing:
                    result["adminApis"].append({"id": identifier, "title": title, "description": description, "path": path, "methods": methods, "supportsCustomUi": False})
        result.update(protocolVersion=2, views=views, capabilities=list(dict.fromkeys(capabilities)), supportedRenderers=list(dict.fromkeys(renderers)), supportedRenderModes=list(dict.fromkeys(modes)), apiContracts=contracts)
        return cast(PluginManifest, parse("PluginManifestSchema", result))

    def schema(self, declaration: ManifestDeclarationInput) -> BpSchemaOutput:
        result = {"manifest": self.manifest(declaration), "routes": [
            {"viewId": route.view_id, "path": route.paths[0], "pathVariants": list(route.paths) if len(route.paths) > 1 else [],
             "operations": [{"operationId": operation.id, "method": operation.method} for operation in route.operations],
             "paramNames": route.param_names, **self._rendering(route)} for route in self.routes]}
        return cast(BpSchemaOutput, parse("BpSchemaOutputSchema", result))

    @staticmethod
    def _rendering(route: Route) -> dict[str, Any]:
        themes, components = [], []
        fragments: dict[tuple[str, str, str, str], dict[str, Any]] = {}
        for operation in route.operations:
            for renderer in operation.handler.renderers:
                theme, kind, key, status = renderer.identity
                if status != 200: continue
                themes.append(theme)
                if kind == "component": components.append(key)
                if kind == "fragment":
                    location, identifier = cast(str, key).split(".")
                    item = fragments.setdefault((location, identifier, operation.id, operation.method),
                        {"fragmentLocation": location, "fragmentId": identifier, "operationId": operation.id, "method": operation.method, "renderers": []})
                    item["renderers"].append(theme)
        return {"renderers": list(dict.fromkeys(themes)), "hasFragments": bool(fragments), "fragments": list(fragments.values()), "components": list(dict.fromkeys(components))}
